Bottom: ec0f566e075b5f61bd6d4e7c1524a241381056c3
Top:    1992102219a3f186a70f208fe4535c142ad59e26
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-14 00:55:17 +0000

Limit virtual memory size of PCB process to avoid swapping my machine to death!

Sometimes the polygon code exploded...


---

diff --git a/src/polygon.c b/src/polygon.c
index 8360d64..af2cc26 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -104,6 +104,9 @@ dicer output is used for HIDs which cannot render things with holes
 #include <dmalloc.h>
 #endif
 
+/* For getrlimit, setrlimit */
+#include <sys/time.h>
+#include <sys/resource.h>
 
 #undef DEBUG_CIRCSEGS
 
@@ -120,6 +123,8 @@ static double bw_rotate_circle_seg[4];
 void
 polygon_init (void)
 {
+  struct rlimit limit;
+
   double cos_ang = cos (2.0 * M_PI / POLY_CIRC_SEGS_F);
   double sin_ang = sin (2.0 * M_PI / POLY_CIRC_SEGS_F);
 
@@ -128,6 +133,12 @@ polygon_init (void)
 
   bw_rotate_circle_seg[0] =  cos_ang;  bw_rotate_circle_seg[1] =  sin_ang;
   bw_rotate_circle_seg[2] = -sin_ang;  bw_rotate_circle_seg[3] =  cos_ang;
+
+  /* DEBUG - AVOID PCB running the system out of memory! */
+  getrlimit (RLIMIT_AS, &limit);
+  limit.rlim_cur = MIN (limit.rlim_cur, 7000 * 1024 * 1024 /* 7000 GiB limit to virtual memory size */);
+  setrlimit (RLIMIT_AS, &limit);
+
 }
 
 Cardinal
